central/networkgraph/service/service_impl.go

```diff
@@ -661,47 +694,178 @@ func filterFlowsAndMaskScopeAlienDeployments(
+	q, err := search.ParseQuery(query, search.MatchAllIfEmpty())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(errox.InvalidArgs, err.Error())
 	}
 
-	return flowStore.GetExternalFlowsForDeployment(ctx, deploymentId)
+	// Retrieves entities where the cluster ID matches the request cluster OR where the cluster ID is empty indicating global entities.
+	clusterMatch := search.DisjunctionQuery(
+		search.MatchFieldQuery(search.ClusterID.String(), search.ExactMatchString(""), false),
+		search.MatchFieldQuery(search.ClusterID.String(), search.ExactMatchString(clusterId), false),
+	)
+
+	q = search.ConjunctionQuery(q, clusterMatch)
+
+	q, _ = search.FilterQueryWithMap(q, schema.NetworkEntitiesSchema.OptionsMap)
+
+	return s.entityDS.GetEntityByQuery(ctx, q)
 }
 
-func (s *serviceImpl) enrichFlowsWithExternalEntityDetails(ctx context.Context, flows *[]*storage.NetworkFlow) error {
-	// This could probably be implemented in the DB query for better efficiency,
-	// but is simply implemented here for now.
-	for _, flow := range *flows {
-		var toGet string
-		var toSet **storage.NetworkEntityInfo
+// This is similar to addDeploymentFlowsToGraph except it will gather only
+// external flows.
+//
+// The process is:
+// (1) gather deployments in the cluster (filtered by the query)
+// (2) gather external entities (also filtered by the query)
+//
+//	(2.1) if entityId is provided, only one entity is retrieved
+//
+// (3) gather flows that satisfy the following rules:
+//
+//	(3.1) are only external flows (deployment <--> external entity)
+//	(3.2) are associated with one of the deployments from (1)
+//	(3.3) are associated with one of the entities from (2)
+//
+// (4) flows are filtered based on permissions (only deployments are constrained in this way, not entities)
+// (5) flows are sorted by source entity ID
+// (6) entities are sorted by ID
+func (s *serviceImpl) getExternalFlowsAndEntitiesByQuery(ctx context.Context, clusterId, query, entityId string, since *time.Time) ([]*storage.NetworkFlow, []*storage.NetworkEntity, error) {
+	deploymentQuery, scopeQuery, err := networkgraph.GetFilterAndScopeQueries(clusterId, query, &v1.NetworkGraphScope{})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to construct filter and scope queries")
+	}
+
+	count, err := s.deployments.Count(ctx, deploymentQuery)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to count deployments")
+	}
 
-		props := flow.GetProps()
-		if props == nil {
-			continue
+	if count > maxNumberOfDeploymentsInGraphEnv.IntegerSetting() {
+		log.Warnf("Number of deployments is too high to be rendered in Network Graph: %d", count)
+		return nil, nil, errors.Errorf(
+			"number of deployments (%d) exceeds maximum allowed for Network Graph: %d",
+			count,
+			maxNumberOfDeploymentsInGraphEnv.IntegerSetting(),
+		)
+	}
+
+	deployments, err := s.deployments.SearchListDeployments(ctx, deploymentQuery)
```

Comment:
Would this work for masked deployments? My understanding is that `SearchListDeployments` will filter out deployments that the current user does not have permission to see (since we are not elevating the permissions on that call any more).

Reply:
This was to bring these endpoints a bit more in line with the way the network graph is constructed, so the filtering behaviour is intended. We do elevate later to get all the flows in the cluster that match the deployments, and then we filter again as a final step (which might be slightly overkill considering at that point we should have a list of flows communicating between only deployments the user can see and external entities)

Reply:
That make sense. Thank you for clarifying!